### PR TITLE
network: several cleanups

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -218,7 +218,11 @@ def wait_online(self, links_with_operstate, timeout='20s', bool_any=False):
         args = [wait_online_bin, f'--timeout={timeout}'] + [f'--interface={link}' for link in links_with_operstate]
         if bool_any:
             args += ['--any']
-        subprocess.check_call(args)
+        rc = subprocess.call(args)
+        if rc != 0:
+            for link in links_with_operstate:
+                subprocess.call(['networkctl', 'status', link.split(':')[0]])
+        self.assertTrue(rc == 0)
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: I think we should be going away from the antiquated unittest interface, not building more checks that use it. What about the following:
```python
try:
    subprocess.check_call(args)
except subprocess.CalledProcessError:
     for link in links_with_operstate:
            subprocess.call(['networkctl', 'status', link.split(':')[0]])
     raise
```

File: test/test-network/systemd-networkd-tests.py [hunk 2]
```diff
@@ -668,14 +668,17 @@ def test_gre_tunnel(self):
         self.assertNotRegex(output, 'iseq')
         self.assertNotRegex(output, 'oseq')
 
-    @expectedFailureIf_ip6gre_do_not_support_ipv6ll()
+    #@expectedFailureIf_ip6gre_do_not_support_ipv6ll()
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: Remove this completely rather than commenting it out.

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -124,6 +125,9 @@ bool link_ipv4ll_enabled(Link *link) {
         if (STRPTR_IN_SET(link->kind, "vrf", "wireguard", "ipip", "gre", "ip6gre", "ip6tnl", "sit", "vti", "vti6"))
                 return false;
 
+        if (IN_SET(link_get_ipvlan_mode(link), NETDEV_IPVLAN_MODE_L3, NETDEV_IPVLAN_MODE_L3S))
```
Comment (keszybz) [inline src/network/networkd-link.c:139]: > disable IPv4LL for ipvlan with L3 or L3S mode

Can you please add a comment why this change (the whole commit) is being done?

File: src/network/netdev/netdev-gperf.gperf [hunk 4]
```diff
@@ -162,6 +162,7 @@ MACsecReceiveAssociation.Activate,        config_parse_macsec_sa_activate,
 Tun.OneQueue,                             config_parse_bool,                         0,                             offsetof(TunTap, one_queue)
 Tun.MultiQueue,                           config_parse_bool,                         0,                             offsetof(TunTap, multi_queue)
 Tun.PacketInfo,                           config_parse_bool,                         0,                             offsetof(TunTap, packet_info)
+Tun.VNetHeader,                           config_parse_bool,                         0,                             offsetof(TunTap, vnet_hdr)
```
Comment (keszybz) [inline src/network/netdev/netdev-gperf.gperf:165]: The commit title implies that this was just omitted by mistake. But afaics, support was added just for tap in back in f5f07dbf06f6df9d12e092c05ad8cfbe244d203f on purpose. So shouldn't the commit just say "Allow IFF_VNET_HDR to also be set for tun devices"? What is the enabling this?

File: src/network/netdev/tuntap.c [hunk 5]
```diff
@@ -134,10 +134,16 @@ static int tuntap_verify(NetDev *netdev, const char *filename) {
         assert(netdev);
 
         if (netdev->mtu != 0)
-                log_netdev_warning(netdev, "MTU configured for %s, ignoring", netdev_kind_to_string(netdev->kind));
+                log_netdev_warning(netdev,
+                                   "MTUBytes= configured for %s device in %s will be ignored. "
```
Comment (keszybz) [inline src/network/netdev/tuntap.c]: Maybe add a newline after the sentence end, here and in the other message?

File: src/network/netdev/tuntap.c [hunk 6]
```diff
@@ -130,6 +131,37 @@ static void tuntap_done(NetDev *netdev) {
...
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        bool *one_queue = data;
+        int r;
+
+        log_syntax(unit, LOG_WARNING, filename, line, 0,
+                   "OneQueue= is deprecated, and may take no effect on newer kernel.");
```
Comment (keszybz) [inline src/network/netdev/tuntap.c]: We require kernel >= 3.13, so this has no effect on all supported kernels. I think this option should just be ignored. Can we use config_parse_warn_compat with DISABLED_LEGACY instead?

File: src/network/networkd-can.c [hunk 7]
```diff
@@ -198,6 +198,8 @@ static int link_down_handler(sd_netlink *rtnl, sd_netlink_message *m, Link *link
 int link_configure_can(Link *link) {
         int r;
 
+        link_set_state(link, LINK_STATE_CONFIGURING);
```
Comment (keszybz) [inline src/network/networkd-can.c:201]: Commit title: make CAN devices go through LINK_STATE_CONFIGURING

File: test/test-network/systemd-networkd-tests.py [hunk 8]
```diff
@@ -445,10 +445,11 @@ def test_bridge(self):
...
 
-        self.assertEqual('900', self.read_link_attr('bridge99', 'bridge', 'hello_time'))
-        self.assertEqual('900', self.read_link_attr('bridge99', 'bridge', 'max_age'))
-        self.assertEqual('900', self.read_link_attr('bridge99', 'bridge','forward_delay'))
-        self.assertEqual('900', self.read_link_attr('bridge99', 'bridge','ageing_time'))
+        # The tests below depend on environment, as they are determined with clock ticks: sysconf(_SC_CLK_TCK).
+        #self.assertEqual('900', self.read_link_attr('bridge99', 'bridge', 'hello_time'))
+        #self.assertEqual('900', self.read_link_attr('bridge99', 'bridge', 'max_age'))
+        #self.assertEqual('900', self.read_link_attr('bridge99', 'bridge','forward_delay'))
+        #self.assertEqual('900', self.read_link_attr('bridge99', 'bridge','ageing_time'))
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: Wow, kernel interfaces rock a usual :) 

So why not do:
```python3
tick = os.sysconf('SC_CLK_TCK')
hello_time = int(self.read_link_attr('bridge99', 'bridge', 'hello_time')) / tick
self.assertEqual(hello_time, 9)
...
```

File: src/network/networkd-link.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-link.c:139]: Not sure, but IPv4LL fails with `-ENOBUFS` in `sendto()`. @ssahani Do you have any idea?

**Comment by yuwata** [comment]
@keszybz Thank you for the review. Most of your comments are addressed. I have no idea about IPv4LL support on `ipvlan` or `ipvtap`, but currently it fails...

**Comment by yuwata** [comment]
@mrc0mmand Could you run the test again? I've slightly updated the logic of bridge slave detection in `networkd`.

File: src/network/networkd-link.c (hunk 3 above)
Comment (ssahani) [reply src/network/networkd-link.c:139]: Could you enable the kernel with ```CONFIG_IPVLAN_L3S``` and test. It seems to me that it may be disabled. But yes ```L3``` mode I tested it's not supported. 
```
IPV4LL: Picked new IP address 169.254.125.86.
ipvlan99: Adding route: dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a
IPV4ACD: Failed to send ARP probe: No buffer space available
IPV4ACD: STOPPED
```
Kernel code I could not figure out why need to look deeper into it. For now I think we can just write ipv4LL not supported for these 2 mode.

Comment (yuwata) [reply src/network/networkd-link.c:139]: I noticed that L3 and L3S mode does not support ARP. And IPv4LL is implemented with arp-util.c (though I have not followed what it does).

Comment (keszybz) [reply src/network/networkd-link.c:139]: So... if the reason is that it fails in tests, than please say that in the commit message.

> Could you enable the kernel with CONFIG_IPVLAN_L3S and test.

CONFIG_IPVLAN_L3S is automatic, if CONFIG_NETFILTER and CONFIG_IPVLAN are enabled.

Comment (yuwata) [reply src/network/networkd-link.c:139]: IPv4LL requires ARP, but L3 and L3S do not support them, so `sendto()` in `arp_send_packet()` fails. This is not only for tests, but it is required to ipvlan or ipvtap interfaces become `configured` operstate (though IPv4LL is disabled by default, so it may be a mis-configuration by a user). But, anyway, we should kindly disable IPv4LL on such mode.

**Comment by mrc0mmand** [comment]
> @mrc0mmand Could you run the test again? I've slightly updated the logic of bridge slave detection in `networkd`.

The fail is finally gone, thank you! Now there's only the one *unexpected success*:

```
test_ip6gre_tunnel (__main__.NetworkdNetDevTests) ... 
56: ip6gretun99@dummy98: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 2a00:ffde:4567:edde::4987 peer 2001:473:fece:cafe::5179 promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote 2001:473:fece:cafe::5179 local 2a00:ffde:4567:edde::4987 dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
57: ip6gretun98@dummy98: <NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 :: brd 2001:473:fece:cafe::5179 promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote 2001:473:fece:cafe::5179 local any dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
58: ip6gretun97@dummy98: <NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 2a00:ffde:4567:edde::4987 brd :: promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote any local 2a00:ffde:4567:edde::4987 dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
unexpected success
```
https://ci.centos.org/job/systemd-ci-build/489/artifact/artifacts_Rvm3Cd/vagrant-logs.b5k/vagrant-arch-testsuite.QSc/systemd-networkd-tests.py_0_FAIL.log

**Comment by yuwata** [comment]
@mrc0mmand Please one more time... If still fails, I will drop the unstable test. So, I hope at most two more times for all tests pass :-)

**Comment by mrc0mmand** [comment]
> @mrc0mmand Please one more time... If still fails, I will drop the unstable test. So, I hope at most two more times for all tests pass :-)

Still the same issue, unfortunately:

```
test_ip6gre_tunnel (__main__.NetworkdNetDevTests) ... 
56: ip6gretun99@dummy98: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 2a00:ffde:4567:edde::4987 peer 2001:473:fece:cafe::5179 promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote 2001:473:fece:cafe::5179 local 2a00:ffde:4567:edde::4987 dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
57: ip6gretun98@dummy98: <NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 :: brd 2001:473:fece:cafe::5179 promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote 2001:473:fece:cafe::5179 local any dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
58: ip6gretun97@dummy98: <NOARP,UP,LOWER_UP> mtu 1448 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/gre6 2a00:ffde:4567:edde::4987 brd :: promiscuity 0 minmtu 0 maxmtu 0 
    ip6gre remote any local 2a00:ffde:4567:edde::4987 dev dummy98 hoplimit 64 encaplimit 0 tclass 0x00 flowlabel 0x00000 addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
unexpected success
```

**Comment by yuwata** [comment]
@mrc0mmand Please test one more time. I hope this is the last trial.

File: man/systemd.netdev.xml [hunk 9]
```diff
@@ -1379,15 +1379,6 @@
     keys:</para>
 
     <variablelist class='network-directives'>
-      <varlistentry>
-        <term><varname>OneQueue=</varname></term>
```
Comment (ssahani) [inline man/systemd.netdev.xml:1383]: Why this is deleted ?

File: src/network/netdev/netdev-gperf.gperf [hunk 10]
```diff
@@ -159,12 +159,13 @@ MACsecReceiveAssociation.KeyId,           config_parse_macsec_key_id,
 MACsecReceiveAssociation.Key,             config_parse_macsec_key,                   0,                             0
 MACsecReceiveAssociation.KeyFile,         config_parse_macsec_key_file,              0,                             0
 MACsecReceiveAssociation.Activate,        config_parse_macsec_sa_activate,           0,                             0
-Tun.OneQueue,                             config_parse_bool,                         0,                             offsetof(TunTap, one_queue)
+Tun.OneQueue,                             config_parse_warn_compat,                  DISABLED_LEGACY,               0
 Tun.MultiQueue,                           config_parse_bool,                         0,                             offsetof(TunTap, multi_queue)
```
Comment (ssahani) [inline src/network/netdev/netdev-gperf.gperf:163]: ahh right kernel have removed it.

**Comment by mrc0mmand** [comment]
> @mrc0mmand Please test one more time. I hope this is the last trial.

Eight iterations in, no fail yet, looks promising!

**Comment by mrc0mmand** [comment]
@yuwata I let it run for at least one *full* run, i.e. 20+ iterations until the machine got garbage collected, and everything passed, so I'd say this PR is ready to go. Thanks a lot for all the effort put into this :-)

**Comment by yuwata** [comment]
@mrc0mmand Thanks a lot!
